File: src/main/java/me/viniciusarnhold/altaria/utils/Actions.java
package me.viniciusarnhold.altaria.utils;

import me.viniciusarnhold.altaria.core.BotManager;
import org.jetbrains.annotations.NotNull;
import sx.blah.discord.util.*;



/**
 * Created by Vinicius.
 *
 * @since ${PROJECT_VERSION}
 */
public final class Actions {

    private Actions() {

    }


    @NotNull
    public static RequestBuilder.IRequestAction ofSuccess(@NotNull IWrappedVoidRequestAction action) throws RateLimitException, MissingPermissionsException, DiscordException {
        return () -> {
            action.execute();
            return true;
        };
    }

    @NotNull
    public static RequestBuilder.IRequestAction ofFailure(@NotNull IWrappedVoidRequestAction action) throws RateLimitException, MissingPermissionsException, DiscordException {
        return () -> {
            action.execute();
            return false;
        };
    }

    public static RequestBuilder.IRequestAction ofSuccess() throws RateLimitException, MissingPermissionsException, DiscordException {
        return () -> true;
    }

    public static RequestBuilder.IRequestAction ofFailure() throws RateLimitException, MissingPermissionsException, DiscordException {
        return () -> false;
    }

    public static <T> RequestBuffer.IRequest<T> wrap(@NotNull IWrappedRequestAction<T> supplier) {
        return () -> {
            try {
                return supplier.request();
            } catch (@NotNull DiscordException | MissingPermissionsException e) {
                BotManager.LOGGER.error(e);
                throw new RuntimeException(e);
            }
        };
    }

    public static RequestBuffer.IVoidRequest wrap(@NotNull IWrappedVoidRequestAction supplier) {
        return () -> {
            try {
                supplier.execute();
            } catch (@NotNull DiscordException | MissingPermissionsException e) {
                BotManager.LOGGER.error(e);
                throw new RuntimeException(e);
            }
        };
    }

    @FunctionalInterface
    public interface IWrappedVoidRequestAction {

        void execute() throws RateLimitException, MissingPermissionsException, DiscordException;

    }

    public interface IWrappedRequestAction<T> {


        @NotNull
        T request() throws RateLimitException, MissingPermissionsException, DiscordException;

    }
}

File: core/src/main/kotlin/me/viniciusarnhold/altaria/utils/Actions.kt
package me.viniciusarnhold.altaria.utils

import me.viniciusarnhold.altaria.core.BotManager
import sx.blah.discord.util.DiscordException
import sx.blah.discord.util.MissingPermissionsException
import sx.blah.discord.util.RateLimitException
import sx.blah.discord.util.RequestBuffer.IRequest
import sx.blah.discord.util.RequestBuffer.IVoidRequest
import sx.blah.discord.util.RequestBuilder.IRequestAction


/**
 * Created by Vinicius.

 * @since ${PROJECT_VERSION}
 */
object Actions {


    @Throws(RateLimitException::class, MissingPermissionsException::class, DiscordException::class)
    fun ofSuccess(action: () -> Unit): () -> Boolean = {
        action()
        true
    }

    @Throws(RateLimitException::class, MissingPermissionsException::class, DiscordException::class)
    fun ofFailure(action: () -> Unit): IRequestAction = IRequestAction {
        action()
        false
    }

    @Throws(RateLimitException::class, MissingPermissionsException::class, DiscordException::class)
    fun ofSuccess(): IRequestAction = IRequestAction { true }

    @Throws(RateLimitException::class, MissingPermissionsException::class, DiscordException::class)
    fun ofFailure(): IRequestAction = IRequestAction { false }

    fun <T> wrap(supplier: () -> T): IRequest<T> = IRequest {
        try {
            supplier()
        } catch (e: DiscordException) {
            BotManager.LOGGER.error(e)
            throw RuntimeException(e)
        } catch (e: MissingPermissionsException) {
            BotManager.LOGGER.error(e)
            throw RuntimeException(e)
        }
    }

    fun wrap(supplier: () -> Unit): IVoidRequest = IVoidRequest {
        try {
            supplier()
        } catch (e: DiscordException) {
            BotManager.LOGGER.error(e)
            throw RuntimeException(e)
        } catch (e: MissingPermissionsException) {
            BotManager.LOGGER.error(e)
            throw RuntimeException(e)
        }
    }
}
